Add BLAKE3 and xxHash (XXH3/XXH3-64 and XXH128/XXH3-128)

albertony opened the pull request:
#### What is the purpose of this change?

Add support for BLAKE3 and xxHash (XXH3/XXH3-64 and XXH128/XXH3-128).
- Added to the rclone's hash library, with internal names: `blake3`, `xxh3` and `xxh128`.
- Implicitly supported in `rclone hashsum` command.
- Implicitly supported by the local backend.
- Added local backend option `--hashes` to configure which hashes it should support.
  - E.g. `:local,hashes=blake3:C:\Temp`.
  - This can be used to test hashing performance of the different types of hashes, e.g. with local-to-local operations.
    - Because the default, as before, is for it to support all that rclone's hash library supports, and when needing one it picks the first of a the internally ordered list, currently `md5`, and with operations between two backends it will pick the first in common. For local-local operations this means it will always use `md5`.
- Added support in SFTP backend.
  - Initially I added only support for BLAKE3: Probes for `blake3sum` executable, or `rclone hashsum blake3`, configurable with `blake3sum_command`.
  - Later this PR was extended to also include other hash algorithms already supported by rclone: `crc32`, `sha256`, `blake3`, `xxh3`, `xxh128`
  - Finally, an option `hashes` to configure which hashes to support was also added (similar to the local backend).
    - This was done due to concerns with startup times, if the backend probes for all supported hashes each time it is used. It therefore only considers `md5` and `sha1` by default, but one can set an option `hashes` to comma separated list of additional/different hashes to support - including crc32, sha256 and the new blake3, xxh3, xxh128. See [comment 1](https://github.com/rclone/rclone/issues/7765#issuecomment-2057057464) and [comment 2](https://github.com/rclone/rclone/pull/7767#issuecomment-2276606131) below.
- Added support in `rclone serve sftp` (edit Dec 2, 2024).

To experiment, beta builds are here: https://beta.rclone.org/branch/add-xxh-blake-hash/

TODO:
- Support for XXH3 in sftp backend
  - The xxhsum tool uses a non-standard prefix "XXH3_" preceding the hash output for the 64-bit variant of XXH3, to avoid confusion with the older 64-bit algorithm XXH64. This was introduced in version 0.8.3 released Dec 30, 2024. Older versions only supported the alternative BSD style output format, otherwise optional with argument --tag. We are currently not expecting these output formats and can therefore not use the "xxhsum -H3" command or its xxh3sum alias directly.
- Name and type of options to specify hash types:
  - Name of the new option for the local and sftp backends introduced here is `hashes`, and it takes a comma-separated list.
  - The hasher backend already has a similar option, with same name `hashes`, and also takes a comma-separated list.
  - The lsjson command has option `hash` to enable hash in output, and then a `hash-type` option, which can be repeated, to specify which ones.
  - The lsf command has option `hash` to specify which hash to use when the option `format` includes keyword "h" enabling hash as part of the output.
  
#### Was the change discussed in an issue or in the forum before?

https://github.com/rclone/rclone/issues/7765 and https://forum.rclone.org/t/faster-non-cryptographic-hashing-algorithm-for-faster-file-comparison/23601

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [ ] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [ ] I'm done, this Pull Request is ready for review :-)

albertony commented:
Rebased and fixed conflicts

albertony commented on backend/local/local.go:
```diff
@@ -270,6 +270,11 @@ only useful for reading.
 				Value: cTime.String(),
 				Help:  "The last status change time.",
 			}},
+		}, {
```
I agree. However, rclone will always use md5 if supported by the other backend, or sha1 etc. These new hashes will therefore never be used, unless they are the only ones the backend of the opposite side of a operation supports (or at least that it does not support the ones rclone prefers first). This means for local-local transfer md5 will always be used. For a SFTP remote one would have to make sure md5 and sha1 are disabled/unavailable if one wants to use the new hashes for a local-sftp transfer. This new config allows me to run a local-local copy operation and force rclone to use XXH3.

For me it was very useful for testing. Maybe not so relevant to include into master, and a more generic solution for hash preferences for backends supporting more than one would probably be better. I included it, however, because I was thinking the new hashes would otherwise have very limited use - I think only restricted to a sftp remote configured to support it and not md5/sha1.

ncw replied:
Ah, I see. Maybe a more generic solution like `--disable-hash MD5,SHA1` might work which could disable the hashes in `lib/hash`. Or maybe something like `--restrict-hash SHA1,XXH3`

albertony commented on backend/sftp/sftp.go:
```diff
@@ -226,6 +226,11 @@ E.g. the second example above should be rewritten as:
 			Help:     "The command used to read sha1 hashes.\n\nLeave blank for autodetect.",
 			Advanced: true,
+		}, {
+			Name:     "blake3sum_command",
```
Yes, I agree, and have noted your concern from an older comment on the TODO list :)

> I have a slight concern about sftp startup times. Lots of people use sftp without a config file which means that it probes for shells/supported hashes each time it is used. Perhaps we should delay hash support probing until it is asked for?
https://github.com/rclone/rclone/issues/7765#issuecomment-2057057464

I was briefly looking into the suggestion of delaying hash support probing, but found that it probably requires somewhat bigger refactoring in core parts that I did not want to dive into... If you mean adding a sftp-specific configuration parameter for enabling additional (or perhaps overriding default) hash types, more like the one I added for the local remote, then that would definitely be easier.

albertony replied:
> Also why blake3? It lost out to keccak in the SHA-3 contest.

I think because that was what was explicitly asked for in the issue, and also one I was familiar with, and know is in use (e.g. CCache uses both XXH3 and BLAKE3). Never heard of  Keccak, I must admit.

ncw replied:
> Yes, I agree, and have noted your concern from an older comment on the TODO list :)

Getting old and repeating myself ;-)

> > I have a slight concern about sftp startup times. Lots of people use sftp without a config file which means that it probes for shells/supported hashes each time it is used. Perhaps we should delay hash support probing until it is asked for?
> > [#7765 (comment)](https://github.com/rclone/rclone/issues/7765#issuecomment-2057057464)
> 
> I was briefly looking into the suggestion of delaying hash support probing, but found that it probably requires somewhat bigger refactoring in core parts that I did not want to dive into...

We already delay hash probing until `Fs.Hashes` is called which is good and doing anything else would require major surgery as you've noted.

Perhaps that is enough?

> If you mean adding a sftp-specific configuration parameter for enabling additional (or perhaps overriding default) hash types, more like the one I added for the local remote, then that would definitely be easier.

That might be a good idea.

Though we should probably inject some science into this - how long does the hash probing take?

```
$ rclone backend --log-format "microseconds" -vv features :sftp,host=localhost,user=sftptest: | jq .Hashes
11:21:56.689201 DEBUG : rclone: Version "v1.68.0-beta.8161.7194d828d.fix-4632-gofile" starting with parameters ["rclone" "backend" "--log-format" "microseconds" "-vv" "features" ":sftp,host=localhost,user=sftptest:"]
11:21:56.691707 DEBUG : Using config file from "/home/ncw/.rclone.conf"
11:21:56.740716 DEBUG : sftp://sftptest@localhost:22/: New connection 127.0.0.1:49344->127.0.0.1:22 to "SSH-2.0-OpenSSH_8.9p1 Ubuntu-3ubuntu0.10"
11:21:57.083580 DEBUG : sftp://sftptest@localhost:22/: Running shell type detection remote command: echo ${ShellId}%ComSpec%
11:21:57.087933 DEBUG : sftp://sftptest@localhost:22/: Remote command result: %ComSpec%
11:21:57.087952 DEBUG : sftp://sftptest@localhost:22/: Shell type "unix" detected (set option shell_type to override)
11:21:57.087978 DEBUG : Saving config "shell_type" in section ":sftp" of the config file
11:21:57.088001 NOTICE: Can't save config "shell_type" for on the fly backend ":sftp"
11:21:57.097247 DEBUG : sftp://sftptest@localhost:22/: Relative path resolved to "/home/sftptest"
11:21:57.097263 DEBUG : sftp://sftptest@localhost:22/: Using root directory "/home/sftptest"
11:21:57.097302 DEBUG : sftp://sftptest@localhost:22/: Checking default md5 hash commands
11:21:57.097483 DEBUG : sftp://sftptest@localhost:22/: Running remote command: md5sum
11:21:57.102713 DEBUG : sftp://sftptest@localhost:22/: Remote command result: d41d8cd98f00b204e9800998ecf8427e  -
11:21:57.102740 DEBUG : sftp://sftptest@localhost:22/: Hash command accepted
11:21:57.102750 DEBUG : sftp://sftptest@localhost:22/: Checking default sha1 hash commands
11:21:57.102919 DEBUG : sftp://sftptest@localhost:22/: Running remote command: sha1sum
11:21:57.107854 DEBUG : sftp://sftptest@localhost:22/: Remote command result: da39a3ee5e6b4b0d3255bfef95601890afd80709  -
11:21:57.107885 DEBUG : sftp://sftptest@localhost:22/: Hash command accepted
11:21:57.107905 DEBUG : sftp://sftptest@localhost:22/: Setting hash command for md5 to "md5sum" (set sha1sum_command to override)
11:21:57.107921 DEBUG : Saving config "md5sum_command" in section ":sftp" of the config file
11:21:57.107945 NOTICE: Can't save config "md5sum_command" for on the fly backend ":sftp"
11:21:57.111972 DEBUG : sftp://sftptest@localhost:22/: Setting hash command for sha1 to "sha1sum" (set md5sum_command to override)
11:21:57.111991 DEBUG : Saving config "sha1sum_command" in section ":sftp" of the config file
11:21:57.112012 NOTICE: Can't save config "sha1sum_command" for on the fly backend ":sftp"
11:21:57.117542 DEBUG : 13 go routines active
[
  "md5",
  "sha1"
]
```

vs

```
$ rclone backend --log-format "microseconds" -vv features :sftp,host=localhost,user=sftptest,disable_hashcheck: | jq .Hashes
11:23:14.557982 DEBUG : rclone: Version "v1.68.0-beta.8161.7194d828d.fix-4632-gofile" starting with parameters ["rclone" "backend" "--log-format" "microseconds" "-vv" "features" ":sftp,host=localhost,user=sftptest,disable_hashcheck:"]
11:23:14.560598 DEBUG : Using config file from "/home/ncw/.rclone.conf"
11:23:14.607004 DEBUG : sftp://sftptest@localhost:22/: New connection 127.0.0.1:52498->127.0.0.1:22 to "SSH-2.0-OpenSSH_8.9p1 Ubuntu-3ubuntu0.10"
11:23:14.963461 DEBUG : sftp://sftptest@localhost:22/: Running shell type detection remote command: echo ${ShellId}%ComSpec%
11:23:14.967806 DEBUG : sftp://sftptest@localhost:22/: Remote command result: %ComSpec%
11:23:14.967821 DEBUG : sftp://sftptest@localhost:22/: Shell type "unix" detected (set option shell_type to override)
11:23:14.967835 DEBUG : Saving config "shell_type" in section ":sftp" of the config file
11:23:14.967857 NOTICE: Can't save config "shell_type" for on the fly backend ":sftp"
11:23:14.977497 DEBUG : sftp://sftptest@localhost:22/: Relative path resolved to "/home/sftptest"
11:23:14.977514 DEBUG : sftp://sftptest@localhost:22/: Using root directory "/home/sftptest"
11:23:14.977763 DEBUG : 13 go routines active
[]
```

So checking 2 hashes went from 419 ms without to 425 ms with on a local machine so 3 ms per hash

The same test on a remote machine gives 703 ms vs 957 ms which is probably more typical given the round trip times. So roughly 120 ms per hash checked. This server is 40 ms ping away so indicating that is 3 round trips.

Could we detect this quicker? Maybe we could do something like

```
$ which md5sum md5 rclone sha1sum sha
/usr/bin/md5sum
/home/ncw/go/bin//rclone
/usr/bin/sha1sum
```

To detect whether all the binaries exist at once? And if the binary exists don't bother to test it. Or only test the ones which actually existed.

Or we could send all the test commands at once, effectively sending a shell script something like

```
$ ssh sftptest@localhost "(md5sum ; md5 ; sha1sum ; sha1 ; rclone md5sum ; rclone sha1sum )</dev/null"
d41d8cd98f00b204e9800998ecf8427e  -
bash: line 1: md5: command not found
da39a3ee5e6b4b0d3255bfef95601890afd80709  -
bash: line 1: sha1: command not found
bash: line 1: rclone: command not found
bash: line 1: rclone: command not found
```

(I think the `</dev/null` is implicit in the way we run these commands).

albertony commented:
WIP: Pushed a new version where only md5 and sha1 are considered by default, but where one can set option `hashes` to comma separated list of additional/different hashes to support - including crc32, sha256 and the new blake3, xxh3, xxh128. Then there are new options for the command line for each of them etc, so it bloats the code a bit, but maybe not too bad... This is kind of in progress still; I want to revisit the code once more, and the functionality is also totally untested as of now, but any thoughts on the general solution are welcome. :)

ncw reviewed (approved):
I think this is looking good now.

Merge for v1.68?

albertony commented:
We *could*.. but its not been through very thorough testing. I was planning on looking through it and test it a bit more myself.

When is the expected deadline for making it in the release?

It is probably not the most anticipated feature, so we could consider just waiting with it until after the release, to have it in the beta for a while...

ncw commented:
Ok Good plan - lets let this cook for a bit longer! I'll probably make the release tomorrow and things are looking good for that :-)

ncw commented:
What do you think about merge for 1.69?

ncw commented:
Great :-) Give me a nudge here when you want me to take a look.

albertony commented:
@ncw you may take another look now.

I must admit I'm having problems testing this, as my sftp test server setup previously used is currently broken, and I'm not sure I'll get time the next few weeks to investigate it further..

Though I did add a new feature, which I have used for some testing: Extend serve sftp from only supporting md5 and sha1 to the new complete set of hashes.

albertony commented:
Rebased on master and fixed merge conflicts.

> @albertony any updates? Do you need some help with testing?

Sure, feel free to try it out! Can you build from the branch, or do you want me to produce a beta build?

ncw reviewed (approved):
This looks ready to merge to me :-)

Please merge when happy @albertony 

Thank you

albertony commented:
I will. I just wanted to do a quick recap of the notes, regarding name and type of options. And also quick look into the new XXH3 format, if the additional changes needed to be able to support it would be very easy to add (but that can be done in a later PR anyway, so not important). Will try to get the time to do it within a few days.
